parsely/src/main/java/com/parsely/parselyandroid/EngagementManager.java

```diff
@@ -0,0 +1,114 @@
+package com.parsely.parselyandroid;
+
+import java.util.Calendar;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.TimeZone;
+import java.util.Timer;
+import java.util.TimerTask;
+
+/**
+ * Engagement manager for article and video engagement.
+ * <p>
+ * Implemented to handle its own queuing of future executions to accomplish
+ * two things:
+ * <p>
+ * 1. Flushing any engaged time before canceling.
+ * 2. Progressive backoff for long engagements to save data.
+ */
+class EngagementManager {
+
+    private final ParselyTracker parselyTracker;
+    public Map<String, Object> baseEvent;
+    private boolean started;
+    private final Timer parentTimer;
+    private TimerTask waitingTimerTask;
```

Comment:
Suggestion (💡): I suggest making this `@Nullable` and then checking that during `stop()`:

```java
    public void stop() {
        if (waitingTimerTask != null) {
            waitingTimerTask.cancel();
        }
        started = false;
    }
```

PS: Actually, I would even recommend for such refactor/test PRs to add any missing nullability annotation wherever possible going forward. 🤷